ogcapi-services/src/error.rs

```diff
@@ -53,35 +61,67 @@ impl Error {
 /// to the client.
 impl IntoResponse for Error {
     fn into_response(self) -> Response {
-        let (status, message) = match self {
+        let status = self.status_code();
+        let exception = Exception::from(self);
+
+        let mut headers = HeaderMap::new();
+        headers.insert(CONTENT_TYPE, PROBLEM_JSON.parse().unwrap());
+
+        (status, headers, Json(exception)).into_response()
+    }
+}
+
+impl From<Error> for Exception {
+    fn from(value: Error) -> Self {
+        let (status, message) = match value {
             // Self::Sqlx(ref e) => {
             //     tracing::error!("SQLx error: {:?}", e);
             //     (self.status_code(), self.to_string())
             // }
-            Self::NotFound => (self.status_code(), self.to_string()),
-            Self::Anyhow(ref e) => {
+            Error::NotFound => (value.status_code(), value.to_string()),
+            Error::Anyhow(ref e) => {
                 tracing::error!("Generic error: {:?}", e);
-                (self.status_code(), self.to_string())
+                (value.status_code(), e.to_string())
             }
-            Self::Url(ref e) => {
+            Error::Url(ref e) => {
                 tracing::error!("Url error: {:?}", e);
-                (self.status_code(), self.to_string())
+                (value.status_code(), e.to_string())
             }
-            Self::Qs(ref e) => {
+            Error::Qs(ref e) => {
                 tracing::error!("Query string error: {:?}", e);
-                (self.status_code(), self.to_string())
+                (value.status_code(), e.to_string())
             }
-            Self::Exception(status, message) => {
+            Error::Exception(status, message) => {
                 tracing::debug!("OGCAPI exception: {}", message);
                 (status, message)
             }
+            Error::OgcApiException(exception) => {
+                return exception;
+            }
         };
 
-        let exception = Exception::new(status.as_u16()).detail(message);
+        Exception::new(status.as_u16()).detail(message)
+    }
+}
 
-        let mut headers = HeaderMap::new();
-        headers.insert(CONTENT_TYPE, PROBLEM_JSON.parse().unwrap());
+/// Helper function to read-lock a RwLock, recovering from poisoning if necessary.
+pub fn read_lock<T>(mutex: &std::sync::RwLock<T>) -> std::sync::RwLockReadGuard<'_, T> {
```

Comment:
Maybe this can be moved to the processes routes, as it is only used there for now. Or the state, if it is intended to be adopted more.